File: apps/server/agentcore/conversation/inline_body.py
"""Inline attachment / mention markers in user ``content``.

Composer pills persist as U+FFFC-delimited tokens so body order is one sequence:
text, materials, and role mentions. Indices address ``attachments[]`` /
``agent_mentions[]`` in appearance order. No extra wire field.

Token: ``\\uFFFC`` + ``A``|``M`` + decimal index + ``\\uFFFC``.
"""

from __future__ import annotations

import re
from collections.abc import Mapping, Sequence
from typing import Literal

OBJECT = "\uFFFC"
TOKEN_RE = re.compile(OBJECT + r"([AM])(\d+)" + OBJECT)

SpanKind = Literal["text", "attachment", "mention"]


def token(kind: Literal["A", "M"], index: int) -> str:
    return f"{OBJECT}{kind}{index}{OBJECT}"


def has_inline_markers(content: str | None) -> bool:
    if not content:
        return False
    return OBJECT in content and TOKEN_RE.search(content) is not None


def parse_inline_body(content: str) -> list[tuple[SpanKind, str | int]]:
    """Return ``(kind, text|index)`` spans. Unknown / malformed tokens stay as text."""
    if not content:
        return []
    spans: list[tuple[SpanKind, str | int]] = []
    pos = 0
    for m in TOKEN_RE.finditer(content):
        if m.start() > pos:
            spans.append(("text", content[pos : m.start()]))
        kind: SpanKind = "attachment" if m.group(1) == "A" else "mention"
        spans.append((kind, int(m.group(2))))
        pos = m.end()
    if pos < len(content):
        spans.append(("text", content[pos:]))
    return spans


def serialize_inline_body(spans: Sequence[tuple[SpanKind, str | int]]) -> str:
    parts: list[str] = []
    for kind, payload in spans:
        if kind == "text":
            parts.append(str(payload).replace(OBJECT, ""))
        elif kind == "attachment":
            parts.append(token("A", int(payload)))
        else:
            parts.append(token("M", int(payload)))
    return "".join(parts)


def plain_text(content: str | None) -> str:
    """User-visible words only — markers (and unknown FFFC) dropped."""
    if not content:
        return ""
    return TOKEN_RE.sub("", content).replace(OBJECT, "")


def migrate_legacy_draft(
    content: str,
    attachment_count: int,
    mention_count: int,
) -> str:
    """Old drafts kept chips outside the body. Append tokens once, at the end."""
    if has_inline_markers(content):
        return content
    if attachment_count <= 0 and mention_count <= 0:
        return content
    extra = "".join(token("A", i) for i in range(attachment_count)) + "".join(
        token("M", i) for i in range(mention_count)
    )
    return content + extra


def _name_of(att: Mapping[str, object] | object, fallback: str) -> str:
    if isinstance(att, Mapping):
        raw = att.get("name") or att.get("path") or fallback
        return str(raw).strip() or fallback
    name = getattr(att, "name", None)
    path = getattr(att, "path", None)
    raw = name or path or fallback
    return str(raw).strip() or fallback


def _kind_of(att: Mapping[str, object] | object) -> str:
    if isinstance(att, Mapping):
        return str(att.get("kind") or "file")
    return str(getattr(att, "kind", None) or "file")


def _role_of(mention: Mapping[str, object] | object, fallback: str) -> str:
    if isinstance(mention, Mapping):
        raw = mention.get("role") or mention.get("agent_id") or fallback
        return str(raw).strip() or fallback
    role = getattr(mention, "role", None)
    agent_id = getattr(mention, "agent_id", None)
    raw = role or agent_id or fallback
    return str(raw).strip() or fallback


_KIND_LABEL = {"file": "文件", "dir": "文件夹", "conversation": "对话"}


def render_inline_labels(
    content: str,
    attachments: Sequence[Mapping[str, object] | object] | None,
    mentions: Sequence[Mapping[str, object] | object] | None,
) -> str:
    """History / title / preview: keep order, don't re-inject file bodies."""
    atts = list(attachments or [])
    ments = list(mentions or [])
    parts: list[str] = []
    for kind, payload in parse_inline_body(content):
        if kind == "text":
            parts.append(str(payload))
            continue
        if kind == "attachment":
            idx = int(payload)
            if 0 <= idx < len(atts):
                att = atts[idx]
                label = _KIND_LABEL.get(_kind_of(att), "文件")
                parts.append(f"[{label} {_name_of(att, str(idx))}]")
            continue
        idx = int(payload)
        if 0 <= idx < len(ments):
            parts.append(f"[点名 {_role_of(ments[idx], str(idx))}]")
    return "".join(parts)


def mention_inline_stub(mention: Mapping[str, object] | object) -> str:
    return f"（点名 {_role_of(mention, '?')}）"


def weave_inline_body(
    content: str,
    file_blocks: Sequence[str],
    mention_stubs: Sequence[str],
) -> str | None:
    """This-turn user message: file bodies sit where the user placed the pills.

    Returns None when there are no markers so callers keep the old envelope path.
    """
    if not has_inline_markers(content):
        return None
    parts: list[str] = []
    for kind, payload in parse_inline_body(content):
        if kind == "text":
            parts.append(str(payload))
            continue
        if kind == "attachment":
            idx = int(payload)
            if 0 <= idx < len(file_blocks):
                block = file_blocks[idx]
                if block:
                    parts.append(block if parts and parts[-1].endswith("\n") else f"\n{block}\n")
            continue
        idx = int(payload)
        if 0 <= idx < len(mention_stubs):
            parts.append(mention_stubs[idx])
    return "".join(parts)


def apply_inline_body(
    user_message: str,
    file_blocks: Sequence[str],
    mentions: Sequence[Mapping[str, object]] | None,
    full_context: str | None,
    slim_context: str | None,
) -> tuple[str, str | None]:
    """If the body has pills, weave materials into it and use the slim envelope."""
    stubs = [mention_inline_stub(m) for m in (mentions or [])]
    woven = weave_inline_body(user_message, file_blocks, stubs)
    if woven is None:
        return user_message, full_context
    return woven, slim_context

File: apps/desktop/src/renderer/lib/inlineBody.ts
/**
 * Inline attachment / mention markers in user ``content``.
 *
 * Composer pills persist as U+FFFC-delimited tokens so body order is one sequence.
 * Indices address attachments[] / agentMentions[] in appearance order.
 * Token: U+FFFC + A|M + decimal index + U+FFFC. Must match
 * ``agentcore.conversation.inline_body``.
 */

export const INLINE_OBJECT = "\uFFFC";

const TOKEN_RE = /\uFFFC([AM])(\d+)\uFFFC/g;

export type InlineSpan =
  | { kind: "text"; text: string }
  | { kind: "attachment"; index: number }
  | { kind: "mention"; index: number };

export function inlineToken(kind: "A" | "M", index: number): string {
  return `${INLINE_OBJECT}${kind}${index}${INLINE_OBJECT}`;
}

const TOKEN_TEST = /\uFFFC[AM]\d+\uFFFC/;

export function hasInlineMarkers(content: string | null | undefined): boolean {
  if (!content) return false;
  return TOKEN_TEST.test(content);
}

export function parseInlineBody(content: string): InlineSpan[] {
  if (!content) return [];
  const spans: InlineSpan[] = [];
  let pos = 0;
  TOKEN_RE.lastIndex = 0;
  let m: RegExpExecArray | null = TOKEN_RE.exec(content);
  while (m) {
    if (m.index > pos) {
      spans.push({ kind: "text", text: content.slice(pos, m.index) });
    }
    const index = Number(m[2]);
    spans.push(
      m[1] === "A" ? { kind: "attachment", index } : { kind: "mention", index },
    );
    pos = m.index + m[0].length;
    m = TOKEN_RE.exec(content);
  }
  if (pos < content.length) {
    spans.push({ kind: "text", text: content.slice(pos) });
  }
  return spans;
}

export function serializeInlineBody(spans: readonly InlineSpan[]): string {
  let out = "";
  for (const span of spans) {
    if (span.kind === "text") {
      out += span.text.replaceAll(INLINE_OBJECT, "");
    } else if (span.kind === "attachment") {
      out += inlineToken("A", span.index);
    } else {
      out += inlineToken("M", span.index);
    }
  }
  return out;
}

export function plainText(content: string | null | undefined): string {
  if (!content) return "";
  return content.replace(TOKEN_RE, "").replaceAll(INLINE_OBJECT, "");
}

export function migrateLegacyDraft(
  content: string,
  attachmentCount: number,
  mentionCount: number,
): string {
  if (hasInlineMarkers(content)) return content;
  if (attachmentCount <= 0 && mentionCount <= 0) return content;
  let extra = "";
  for (let i = 0; i < attachmentCount; i++) extra += inlineToken("A", i);
  for (let i = 0; i < mentionCount; i++) extra += inlineToken("M", i);
  return content + extra;
}

export function insertInlineToken(
  content: string,
  caret: number,
  kind: "A" | "M",
  index: number,
): { value: string; caret: number } {
  const safe = Math.max(0, Math.min(caret, content.length));
  const tok = inlineToken(kind, index);
  return {
    value: content.slice(0, safe) + tok + content.slice(safe),
    caret: safe + tok.length,
  };
}

export function stripRange(
  content: string,
  start: number,
  end: number,
): { value: string; caret: number } {
  const a = Math.max(0, Math.min(start, end, content.length));
  const b = Math.max(0, Math.min(Math.max(start, end), content.length));
  return { value: content.slice(0, a) + content.slice(b), caret: a };
}

export function dropInlineIndex(
  content: string,
  kind: "attachment" | "mention",
  index: number,
): string {
  return serializeInlineBody(
    parseInlineBody(content).flatMap((s) => {
      if (s.kind !== kind) return [s];
      if (s.index === index) return [];
      if (s.index > index) return [{ ...s, index: s.index - 1 }];
      return [s];
    }),
  );
}

export function reconcileInlineBody<
  A extends { id: string },
  M extends { id: string },
>(
  content: string,
  attachments: readonly A[],
  mentions: readonly M[],
): { value: string; attachments: A[]; mentions: M[] } {
  const nextAtts: A[] = [];
  const nextMents: M[] = [];
  const spans: InlineSpan[] = [];
  for (const span of parseInlineBody(content)) {
    if (span.kind === "text") {
      spans.push(span);
      continue;
    }
    if (span.kind === "attachment") {
      const item = attachments[span.index];
      if (!item) continue;
      spans.push({ kind: "attachment", index: nextAtts.length });
      nextAtts.push(item);
      continue;
    }
    const item = mentions[span.index];
    if (!item) continue;
    spans.push({ kind: "mention", index: nextMents.length });
    nextMents.push(item);
  }
  return {
    value: serializeInlineBody(spans),
    attachments: nextAtts,
    mentions: nextMents,
  };
}

const KIND_LABEL: Record<string, string> = {
  file: "文件",
  dir: "文件夹",
  conversation: "对话",
};

function nameOf(
  item: { name?: string; path?: string } | undefined,
  fallback: string,
): string {
  const raw = item?.name || item?.path || fallback;
  return raw.trim() || fallback;
}

function roleOf(
  item: { role?: string; agentId?: string; agent_id?: string } | undefined,
  fallback: string,
): string {
  const raw = item?.role || item?.agentId || item?.agent_id || fallback;
  return raw.trim() || fallback;
}

/** History / copy / title preview: keep order, never re-inject file bodies. */
export function renderInlineLabels(
  content: string,
  attachments: ReadonlyArray<{
    name?: string;
    path?: string;
    kind?: string;
  }> | null,
  mentions: ReadonlyArray<{
    role?: string;
    agentId?: string;
    agent_id?: string;
  }> | null,
): string {
  const atts = attachments ?? [];
  const ments = mentions ?? [];
  let out = "";
  for (const span of parseInlineBody(content)) {
    if (span.kind === "text") {
      out += span.text;
      continue;
    }
    if (span.kind === "attachment") {
      const att = atts[span.index];
      if (!att) continue;
      const label = KIND_LABEL[att.kind ?? "file"] ?? "文件";
      out += `[${label} ${nameOf(att, String(span.index))}]`;
      continue;
    }
    const mention = ments[span.index];
    if (!mention) continue;
    out += `[点名 ${roleOf(mention, String(span.index))}]`;
  }
  return out;
}
